src/gt4sd/frameworks/gflownet/dataloader/data_module.py

```diff
@@ -0,0 +1,210 @@
+#
+# MIT License
+#
+# Copyright (c) 2022 GT4SD team
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+"""Data module for gflownet."""
+
+import logging
+from typing import Any, Dict, Optional
+
+import sentencepiece as _sentencepiece
+import numpy as np
+import pytorch_lightning as pl
+import torch.nn as nn
+from torch.utils.data import DataLoader  # , Subset, random_split
+
+from ..dataloader.dataset import GFlowNetDataset, GFlowNetTask
+from ..envs.graph_building_env import GraphBuildingEnv, GraphBuildingEnvContext
+from ..loss.trajectory_balance import TrajectoryBalance
+from ..ml.models import MODEL_FACTORY
+from .sampler import SamplingIterator
+
+# sentencepiece has to be loaded before lightning to avoid segfaults
+_sentencepiece
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+class GFlowNetDataModule(pl.LightningDataModule):
+    """Data module from gflownet."""
+
+    def __init__(
+        self,
+        configuration: Dict[str, Any],
+        dataset: GFlowNetDataset,
+        environment: GraphBuildingEnv,
+        context: GraphBuildingEnvContext,
+        task: GFlowNetTask,
+        algorithm: TrajectoryBalance,
+        model: nn.Module,
+    ) -> None:
+        """Construct GFlowNetDataModule.
+
+        The module assumes a model and algorithm factory/registry.
+        The user should provide a dataset, environment, context for the environment, and task.
+
+        Args:
+            configuration: configuration dictionary.
```

Comment:
I agree in principle. The problem is that we have a lot of hyperparameters in GFlowNet and the class definition in `train/core.py` is confusing if we specify all the hps. Anyway, we can quickly revert this design choice at any stage in the development.

Reply:
👍 
I understand,

In case, it will make sense at some point in the future.
Once you've defined all the parameters in the constructor, you could simply write:
GFlowNetDataModule(**cfg["data_module"])